### PR TITLE
implement arithmetic right shift

### PR DESCRIPTION
**Pull request by jangko**

### CODE REVIEW

**Comment by jangko** [comment]
this PR is ready, review please.
summary:
* implement int version of `shl`
* implement int version of `shr`
* implement int version of `ashr`
* implement both runtime and compiletile version of above functionalities
* fix int version of `high()`'s bug

File: stint/private/int_bitwise_ops.nim [hunk 1]
```diff
@@ -24,3 +24,74 @@ func `and`*(x, y: IntImpl): IntImpl {.inline.}=
...
+  ## than the number of bits in x.
+  const halfSize: type(y) = bitsof(x) div 2
+  type LoType = type(result.lo)
+  if y == 0:
+    return x
+  elif y == halfSize:
+    result.lo = convert[LoType](x.hi)
+    result.hi = ashr(x.hi, halfSize-1)
+  elif y < halfSize:
+    result.lo = (x.lo shr y) or convert[LoType](x.hi shl (halfSize - y))
```
Comment (mratsim) [inline stint/private/int_bitwise_ops.nim]: are you sure we shouldn't `ashr` here?

**Comment by jangko** [comment]
> By the way, I'm curious, what suddenly requires signed bigints? AFAIK all 256-bit integers should be unsigned.

[EIP-145](https://eips.ethereum.org/EIPS/eip-145) requires signed int operation, especially `SAR` opcode.
Implementing signed int bitwise operations for nim-stint also improves the library overall quality. People usually expect bitwise operations available for both int and uint. And missing `SHL`, `SHR`, or `SAR` for signed int could be annoying.

> ideally we should have add testing against ttmath

I don't mind add more tests. In fact during testing compile time `SAR` I discover bug in `high(stint)`. Other bug I discover while testing EVM is the bigEndian version of `initFromBytesBE` failed to compile(but I didn't fix it).

More tests is better, it will increase our confidence.

File: stint/private/int_bitwise_ops.nim (hunk 1 above)
Comment (jangko) [reply stint/private/int_bitwise_ops.nim]: By definition, only `x.hi` part carries the `sign` bit. `x.lo` never have sign. And the `y` is not crossing the `hi-lo` boundary. It is safe to use only `shr` here.

for example, s = sign bit, z = hi bits, u = lo bits:

```text
      szzzzzzzuuuuuuuu

      sar 3                       # hi-lo boundary is 8, y is 3

      sssszzzzzzzuuuuu   # no need to use ashr for lo bits
```

**Comment by jangko** [comment]
it pass tests against ttmath, but when I run `property_based_uint256.nim` why `mod`, `div`, and `pow` not pass the test? undefined behaviour?

**Comment by mratsim** [comment]
sounds good

Strange I'll open a bug for the mod/div/pow it should have disappear after https://github.com/status-im/nim-stint/commit/0987af4467bcd940c978468dfaebb9d77aa13daa, https://github.com/status-im/nim-stint/pull/32,

**Pull request merged**